mlflow/types/schema.py

```diff
@@ -122,34 +126,291 @@ def from_numpy_type(cls, np_type):
+        d["required"] = self.required
+        return {self.name: d}
+
+    @classmethod
+    def from_json_dict(cls, **kwargs):
+        """
+        Deserialize from a json loaded dictionary.
+        The dictionary is expected to contain only one key as `name`, and
+        the value should be a dictionary containing `type` and
+        optional `required` keys.
+        Example: {"property_name": {"type": "string", "required": True}}
+        """
+        if len(kwargs) != 1:
+            raise MlflowException(
+                f"Expected Property JSON to contain a single key as name, got {len(kwargs)} keys."
+            )
+        name, dic = kwargs.popitem()
+        if not {"type"} <= set(dic.keys()):
+            raise MlflowException(f"Missing keys in Property `{name}`. Expected to find key `type`")
+        required = dic.pop("required", True)
+        dtype = dic["type"]
+        if dtype not in [ARRAY_TYPE, OBJECT_TYPE]:
+            return cls(name=name, dtype=dtype, required=required)
+        if dtype == ARRAY_TYPE:
+            return cls(name=name, dtype=Array.from_json_dict(**dic), required=required)
+        if dtype == OBJECT_TYPE:
+            return cls(name=name, dtype=Object.from_json_dict(**dic), required=required)
+
+
+@experimental
+class Object:
+    """
+    Specification used to represent a json-convertible object.
+    """
+
+    def __init__(self, properties: List[Property]) -> None:
+        self._check_properties(properties)
+        self._properties = properties
+
+    def _check_properties(self, properties):
+        if not isinstance(properties, list):
+            raise MlflowException.invalid_parameter_value(
+                f"Expected properties to be a list, got type {type(properties).__name__}"
+            )
+        if len(properties) == 0:
+            raise MlflowException.invalid_parameter_value(
+                "Creating Object with empty properties is not allowed."
+            )
+        if any(not isinstance(v, Property) for v in properties):
+            raise MlflowException.invalid_parameter_value(
+                "Expected values to be instance of Property"
+            )
+        # check duplicated property names
+        names = [prop.name for prop in properties]
+        duplicates = {name for name in names if names.count(name) > 1}
+        if len(duplicates) > 0:
+            raise MlflowException.invalid_parameter_value(
+                f"Found duplicated property names: {duplicates}"
+            )
+
+    @property
+    def properties(self) -> List[Property]:
+        """The list of object properties"""
+        return self._properties
+
+    @properties.setter
+    def properties(self, value: List[Property]) -> None:
+        self._check_properties(value)
+        self._properties = value
+
+    def __eq__(self, other) -> bool:
+        if isinstance(other, Object):
+            return sorted(self.properties, key=lambda x: x.name) == sorted(
```

Comment:
Can we make `Property` sortable? https://stackoverflow.com/questions/7152497/making-a-python-user-defined-class-sortable-hashable